pkgs/build-support/rust/default.nix

```diff
@@ -114,24 +114,39 @@ stdenv.mkDerivation (filteredArgs // {
     EOF
 
     export RUST_LOG=${logLevel}
-  '' + stdenv.lib.optionalString validateCargoDeps ''
-    if ! diff source/Cargo.lock $cargoDepsCopy/Cargo.lock ; then
-      echo
-      echo "ERROR: cargoSha256 is out of date"
-      echo
-      echo "Cargo.lock is not the same in $cargoDepsCopy"
-      echo
-      echo "To fix the issue:"
-      echo '1. Use "1111111111111111111111111111111111111111111111111111" as the cargoSha256 value'
-      echo "2. Build the derivation and wait it to fail with a hash mismatch"
-      echo "3. Copy the 'got: sha256:' value back into the cargoSha256 field"
-      echo
-
-      exit 1
-    fi
+  '' + (args.postUnpack or "");
+
+  # After unpacking and applying patches, check that the Cargo.lock matches our
+  # src package. Note that we do this after the patchPhase, because the
+  # patchPhase may create the Cargo.lock if upstream has not shipped one.
+  postPatch = (args.postPatch or "") + stdenv.lib.optionalString validateCargoDeps ''
+    (
+      cd ..
+
+      # Use a glob to avoid a dependency on the name of the unpacked source dir.
+      cargoDepsLockfile=$cargoDepsCopy/Cargo.lock
+      srcLockfile=$(ls */Cargo.lock | grep -v $cargoDepsLockfile)
```

Comment:
LMK if there's a better way to do this, but this removes the implicit dependency on having come out of `fetchFromGitHub` and having a src directory that is exactly named "source".  E.g., it allows us to delete things like this:

```nix
  # N.B. The cargo vendor consistency checker searches in "source" for lockfile
  cargoDepsHook = ''
    ln -s wishbone-tool source
  '';
```

Reply:
the Cargo.lock should be located at `$sourceRoot/Cargo.lock`. sourceRoot gets set by the unpackPhase if I remember correctly.